### PR TITLE
sim/pi: support I2C

### PR DESCRIPTION
**Pull request by mlyle**
status: hmc5883 tested, px4flow not (have a cbaling issue)

### CODE REVIEW

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
px4flow tested and happy now.  (a bit hacky, but everythign in this area needs help)

File: flight/Modules/Sensors/sensors.c [hunk 1]
```diff
@@ -163,12 +163,14 @@ int32_t SensorsInitialize(void)
 		case OPTICALFLOWSETTINGS_SENSORTYPE_PX4FLOW:
 #if defined(PIOS_INCLUDE_PX4FLOW)
 		{
-			struct pios_px4flow_cfg pios_px4flow_cfg;
-			pios_px4flow_cfg.rotation.roll_D100 = opticalFlowSettings.SensorRotation[OPTICALFLOWSETTINGS_SENSORROTATION_ROLL];
-			pios_px4flow_cfg.rotation.pitch_D100 = opticalFlowSettings.SensorRotation[OPTICALFLOWSETTINGS_SENSORROTATION_PITCH];
-			pios_px4flow_cfg.rotation.yaw_D100 = opticalFlowSettings.SensorRotation[OPTICALFLOWSETTINGS_SENSORROTATION_YAW];
-			if (PIOS_PX4Flow_Init(&pios_px4flow_cfg, external_i2c_adapter_id) != 0) {
-				// set alarm
+			if (external_i2c_adapter_id) {
```
Comment (tracernz) [inline flight/Modules/Sensors/sensors.c:166]: It makes sense to alarm this even without an adaptor on normal targets doesn't it? That is one of the potential causes of PX4 not working despite being selected...

File: flight/PiOS/posix/pios_sys.c [hunk 2]
```diff
@@ -183,6 +204,65 @@ static int handle_serial_device(const char *optarg) {
 }
 #endif
 
+#ifdef PIOS_INCLUDE_I2C
+static int handle_i2c_device(const char *optarg) {
+	char arg_copy[128];
+
+	strncpy(arg_copy, optarg, sizeof(arg_copy));
+	arg_copy[sizeof(arg_copy)-1] = 0;
```
Comment (tracernz) [inline flight/PiOS/posix/pios_sys.c]: Missing spaces. 😛

File: flight/PiOS/posix/pios_sys.c [hunk 3]
```diff
@@ -183,6 +204,65 @@ static int handle_serial_device(const char *optarg) {
...
+
+	char *saveptr;
+
+	char *drv_name = strtok_r(arg_copy, ":", &saveptr);
+	if (drv_name == NULL) goto fail;
```
Comment (tracernz) [inline flight/PiOS/posix/pios_sys.c]: Should be a newline here I feel.

File: flight/PiOS/posix/pios_sys.c [hunk 4]
```diff
@@ -183,6 +204,65 @@ static int handle_serial_device(const char *optarg) {
...
+
+	char *bus_num_str = strtok_r(NULL, ":", &saveptr);
+	if (bus_num_str == NULL) goto fail;
```
Comment (tracernz) [inline flight/PiOS/posix/pios_sys.c]: And here.

File: flight/PiOS/posix/pios_sys.c [hunk 5]
```diff
@@ -239,7 +319,12 @@ static int handle_device(const char *optarg) {
 		bmm150_cfg = PIOS_malloc(sizeof(*bmm150_cfg));
 		bzero(bmm150_cfg, sizeof(*bmm150_cfg));
 
-		bmm150_cfg->orientation = PIOS_BMM_TOP_90DEG;
+		if (mag_orientation == 255) {
+			bmm150_cfg->orientation = PIOS_BMM_TOP_90DEG;
+		} else {
+			/* XXX figure out orientation properly */
```
Comment (tracernz) [inline flight/PiOS/posix/pios_sys.c:329]: Out of scope this time?

**Review by tracernz — COMMENTED** [review]
Cool. 👍

File: flight/Modules/Sensors/sensors.c (hunk 1 above)
Comment (mlyle) [reply flight/Modules/Sensors/sensors.c:166]: This has no chance to work on normal targets right now, I believe.  I'm about to make a systematic path through px4flow to make it do ... somethign better.

File: flight/PiOS/posix/pios_sys.c (hunk 5 above)
Comment (mlyle) [reply flight/PiOS/posix/pios_sys.c:329]: Yah.  This is kinda a weird case (specified an explicit mag orientation, but using the flyingpi's mag).  So I'm not entirely sure what's right.

**Review by tracernz — APPROVED** [review]

**Pull request merged**